File: python-examples/rpa-auth-example/api/get-consultations-by-email.py
from playwright.async_api import Page, BrowserContext
from typing import List
from intuned_browser import go_to_url

from utils.types_and_schemas import GetConsultationsByEmailSchema, Consultation


async def search_by_email(page: Page, email: str):
    search_input = page.locator("input[placeholder='Search by name, email, or phone']")
    await search_input.wait_for(state="visible")
    await search_input.fill(email)

    search_button = page.locator("#search-btn")
    await search_button.click()

    # Wait for search results to load
    await page.wait_for_timeout(1000)


async def find_consultation_items(page: Page):
    consultation_items = await page.locator(".consultation-item").all()
    print(f"Found {len(consultation_items)} consultations")
    return consultation_items


async def extract_consultations_data(consultation_items: list) -> List[Consultation]:
    consultations: List[Consultation] = []

    for consultation_item in consultation_items:
        try:
            # Extract the consultation ID
            id_element = consultation_item.locator(".consultation-id")
            consultation_id = await id_element.text_content()

            # Extract the status
            status_element = consultation_item.locator(".consultation-status")
            status = await status_element.text_content()

            # Extract client name (using .data-value to get clean text)
            client_name_element = consultation_item.locator(".client-name .data-value")
            client_name = await client_name_element.text_content()

            # Extract client email (using .data-value to get clean text)
            email_element = consultation_item.locator(".client-email .data-value")
            email = await email_element.text_content()

            # Extract client phone (using .data-value to get clean text)
            phone_element = consultation_item.locator(".client-phone .data-value")
            phone = await phone_element.text_content()

            # Extract consultation date (using .data-value to get clean text)
            date_element = consultation_item.locator(".consultation-date .data-value")
            date = await date_element.text_content()

            # Extract consultation time (using .data-value to get clean text)
            time_element = consultation_item.locator(".consultation-time .data-value")
            time = await time_element.text_content()

            # Extract consultation topic (using .data-value to get clean text)
            topic_element = consultation_item.locator(".consultation-topic .data-value")
            topic = await topic_element.text_content()

            # Add the consultation to our list
            consultations.append(
                Consultation(
                    id=consultation_id.strip() if consultation_id else "",
                    status=status.strip() if status else "",
                    clientName=client_name.strip() if client_name else "",
                    email=email.strip() if email else "",
                    phone=phone.strip() if phone else "",
                    date=date.strip() if date else "",
                    time=time.strip() if time else "",
                    topic=topic.strip() if topic else "",
                )
            )

            print(f"Successfully extracted consultation: {consultation_id}")
        except Exception as error:
            # If extraction fails for a single consultation, log it but continue with others
            print(f"Failed to extract consultation data: {error}")
            continue

    return consultations


async def automation(
    page: Page,
    params: dict | None = None,
    context: BrowserContext | None = None,
    **_kwargs,
) -> List[Consultation]:
    if params is None:
        raise ValueError("Params are required for this automation")

    # Validate params using pydantic model
    validated_params = GetConsultationsByEmailSchema(**params)

    # Step 1: Navigate to the consultations list page
    # waitUntil: "networkidle" ensures all consultations are loaded
    await go_to_url(
        page=page,
        url="https://sandbox.intuned.dev/consultations-auth/list",
        wait_for_load_state="networkidle",
    )

    # Step 2: Wait for the consultations list container to be visible
    # This ensures the page has loaded before we try to extract data
    consultations_list = page.locator("#consultations-list")
    await consultations_list.wait_for(state="visible")

    # Step 3: Search by email
    await search_by_email(page, validated_params.email)

    # Step 4: Find all consultation items on the page
    # Each consultation item has the class "consultation-item"
    consultation_items = await find_consultation_items(page)

    # Step 5: Extract data from each consultation item
    consultations = await extract_consultations_data(consultation_items)

    # Step 6: Return all extracted consultations
    print(f"Successfully extracted {len(consultations)} consultations")

    return consultations

File: typescript-examples/rpa-example/api/get-consultations-by-email.ts
import { z } from "zod";
import { BrowserContext, Page } from "playwright";
import { goToUrl } from "@intuned/browser";

import {
  getConsultationsByEmailSchema,
  Consultation,
} from "../utils/typesAndSchemas";

type Params = z.infer<typeof getConsultationsByEmailSchema>;

async function searchByEmail(page: Page, email: string) {
  const searchInput = page.locator(
    "input[placeholder='Search by name, email, or phone']"
  );
  await searchInput.waitFor({ state: "visible" });
  await searchInput.fill(email);

  const searchButton = page.locator("#search-btn");
  await searchButton.click();

  // Wait for search results to load
  await page.waitForTimeout(1000);
}

async function findConsultationItems(page: Page) {
  const consultationItems = await page.locator(".consultation-item").all();
  console.log(`Found ${consultationItems.length} consultations`);
  return consultationItems;
}

async function extractConsultationsData(
  consultationItems: any[]
): Promise<Consultation[]> {
  const consultations: Consultation[] = [];

  for (const consultationItem of consultationItems) {
    try {
      // Extract the consultation ID
      const idElement = consultationItem.locator(".consultation-id");
      const id = await idElement.textContent();

      // Extract the status
      const statusElement = consultationItem.locator(".consultation-status");
      const status = await statusElement.textContent();

      // Extract client name (using .data-value to get clean text)
      const clientNameElement = consultationItem.locator(
        ".client-name .data-value"
      );
      const clientName = await clientNameElement.textContent();

      // Extract client email (using .data-value to get clean text)
      const emailElement = consultationItem.locator(
        ".client-email .data-value"
      );
      const email = await emailElement.textContent();

      // Extract client phone (using .data-value to get clean text)
      const phoneElement = consultationItem.locator(
        ".client-phone .data-value"
      );
      const phone = await phoneElement.textContent();

      // Extract consultation date (using .data-value to get clean text)
      const dateElement = consultationItem.locator(
        ".consultation-date .data-value"
      );
      const date = await dateElement.textContent();

      // Extract consultation time (using .data-value to get clean text)
      const timeElement = consultationItem.locator(
        ".consultation-time .data-value"
      );
      const time = await timeElement.textContent();

      // Extract consultation topic (using .data-value to get clean text)
      const topicElement = consultationItem.locator(
        ".consultation-topic .data-value"
      );
      const topic = await topicElement.textContent();

      // Add the consultation to our list
      consultations.push({
        id: id?.trim() || "",
        status: status?.trim() || "",
        clientName: clientName?.trim() || "",
        email: email?.trim() || "",
        phone: phone?.trim() || "",
        date: date?.trim() || "",
        time: time?.trim() || "",
        topic: topic?.trim() || "",
      });

      console.log(`Successfully extracted consultation: ${id}`);
    } catch (error) {
      // If extraction fails for a single consultation, log it but continue with others
      console.error(`Failed to extract consultation data: ${error}`);
      continue;
    }
  }

  return consultations;
}

export default async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
): Promise<Consultation[]> {
  // Validate params using Zod schema
  const { email } = getConsultationsByEmailSchema.parse(params);

  // Step 1: Navigate to the consultations list page
  await goToUrl({
    page: page,
    url: "https://sandbox.intuned.dev/consultations/list",
  });

  // Step 2: Wait for the consultations list container to be visible
  // This ensures the page has loaded before we try to extract data
  const consultationsList = page.locator("#consultations-list");
  await consultationsList.waitFor({ state: "visible" });

  // Step 3: Search by email
  await searchByEmail(page, email);

  // Step 4: Find all consultation items on the page
  // Each consultation item has the class "consultation-item"
  const consultationItems = await findConsultationItems(page);

  // Step 5: Extract data from each consultation item
  const consultations = await extractConsultationsData(consultationItems);

  // Step 6: Return all extracted consultations
  console.log(`Successfully extracted ${consultations.length} consultations`);

  return consultations;
}
